<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>生物信息R数据分析</title>
  <meta name="description" content="生物信息R数据分析">
  <meta name="generator" content="bookdown 0.5 and GitBook 2.6.7">

  <meta property="og:title" content="生物信息R数据分析" />
  <meta property="og:type" content="book" />
  
  <meta property="og:image" content="images/cover.jpg" />
  <meta property="og:description" content="生物信息R数据分析" />
  <meta name="github-repo" content="xie186/HarvardDataScienceForLifeScience_cn" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="生物信息R数据分析" />
  
  <meta name="twitter:description" content="生物信息R数据分析" />
  <meta name="twitter:image" content="images/cover.jpg" />

<meta name="author" content="作者：Rafael A. Irizarry; Mike I. Love 翻译：张三 李四 麻子">


<meta name="date" content="2018-05-05">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="index.html">
<link rel="next" href="inference.html">
<script src="libs/jquery/jquery.min.js"></script>
<link href="libs/gitbook/css/style.css" rel="stylesheet" />
<link href="libs/gitbook/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>

<link rel="stylesheet" href="css/style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">生物信息R数据分析</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Cover picture</a></li>
<li class="chapter" data-level="" data-path="acknowledgments.html"><a href="acknowledgments.html"><i class="fa fa-check"></i>Acknowledgments</a><ul>
<li class="chapter" data-level="0.1" data-path="acknowledgments.html"><a href="acknowledgments.html#section-0.1"><i class="fa fa-check"></i><b>0.1</b> 简介</a></li>
<li class="chapter" data-level="" data-path="acknowledgments.html"><a href="acknowledgments.html#e8bf99e69cace4b9a6e58c85e590abe4bb80e4b988e58685e5aeb9efbc9f"><i class="fa fa-check"></i>这本书包含什么内容？</a></li>
<li class="chapter" data-level="" data-path="acknowledgments.html"><a href="acknowledgments.html#how-is-this-book-different"><i class="fa fa-check"></i>How Is This Book Different?</a></li>
<li class="chapter" data-level="0.2" data-path="acknowledgments.html"><a href="acknowledgments.html#section-0.2"><i class="fa fa-check"></i><b>0.2</b> 这本书有什么不同？</a></li>
<li class="chapter" data-level="0.3" data-path="acknowledgments.html"><a href="acknowledgments.html#r"><i class="fa fa-check"></i><b>0.3</b> 安装R</a></li>
<li class="chapter" data-level="0.4" data-path="acknowledgments.html"><a href="acknowledgments.html#r"><i class="fa fa-check"></i><b>0.4</b> R基础知识</a></li>
<li class="chapter" data-level="0.5" data-path="acknowledgments.html"><a href="acknowledgments.html#installing-packages"><i class="fa fa-check"></i><b>0.5</b> Installing Packages</a></li>
<li class="chapter" data-level="0.6" data-path="acknowledgments.html"><a href="acknowledgments.html#importing-data-into-r"><i class="fa fa-check"></i><b>0.6</b> Importing Data into R</a><ul>
<li class="chapter" data-level="0.6.1" data-path="acknowledgments.html"><a href="acknowledgments.html#getting-started-exercises"><i class="fa fa-check"></i><b>0.6.1</b> Getting Started Exercises</a></li>
</ul></li>
<li class="chapter" data-level="0.7" data-path="acknowledgments.html"><a href="acknowledgments.html#brief-introduction-to-dplyr"><i class="fa fa-check"></i><b>0.7</b> Brief Introduction to <code>dplyr</code></a><ul>
<li class="chapter" data-level="0.7.1" data-path="acknowledgments.html"><a href="acknowledgments.html#dplyr-exercises"><i class="fa fa-check"></i><b>0.7.1</b> <code>dplyr</code> exercises</a></li>
</ul></li>
<li class="chapter" data-level="0.8" data-path="acknowledgments.html"><a href="acknowledgments.html#mathematical-notation"><i class="fa fa-check"></i><b>0.8</b> Mathematical Notation</a></li>
</ul></li>
<li class="chapter" data-level="1" data-path="inference.html"><a href="inference.html"><i class="fa fa-check"></i><b>1</b> Inference</a><ul>
<li class="chapter" data-level="1.1" data-path="inference.html"><a href="inference.html#introduction"><i class="fa fa-check"></i><b>1.1</b> Introduction</a></li>
<li class="chapter" data-level="1.2" data-path="inference.html"><a href="inference.html#random-variables"><i class="fa fa-check"></i><b>1.2</b> Random Variables</a></li>
<li class="chapter" data-level="1.3" data-path="inference.html"><a href="inference.html#the-null-hypothesis"><i class="fa fa-check"></i><b>1.3</b> The Null Hypothesis</a></li>
<li class="chapter" data-level="1.4" data-path="inference.html"><a href="inference.html#distributions"><i class="fa fa-check"></i><b>1.4</b> Distributions</a></li>
<li class="chapter" data-level="1.5" data-path="inference.html"><a href="inference.html#probability-distribution"><i class="fa fa-check"></i><b>1.5</b> Probability Distribution</a></li>
<li class="chapter" data-level="1.6" data-path="inference.html"><a href="inference.html#normal-distribution"><i class="fa fa-check"></i><b>1.6</b> Normal Distribution</a></li>
<li class="chapter" data-level="1.7" data-path="inference.html"><a href="inference.html#populations-samples-and-estimates"><i class="fa fa-check"></i><b>1.7</b> Populations, Samples and Estimates</a><ul>
<li class="chapter" data-level="1.7.1" data-path="inference.html"><a href="inference.html#population-samples-and-estimates-exercises"><i class="fa fa-check"></i><b>1.7.1</b> Population, Samples, and Estimates Exercises</a></li>
</ul></li>
<li class="chapter" data-level="1.8" data-path="inference.html"><a href="inference.html#central-limit-theorem-and-t-distribution"><i class="fa fa-check"></i><b>1.8</b> Central Limit Theorem and t-distribution</a></li>
<li class="chapter" data-level="1.9" data-path="inference.html"><a href="inference.html#central-limit-theorem-in-practice"><i class="fa fa-check"></i><b>1.9</b> Central Limit Theorem in Practice</a></li>
<li class="chapter" data-level="1.10" data-path="inference.html"><a href="inference.html#t-tests-in-practice"><i class="fa fa-check"></i><b>1.10</b> t-tests in Practice</a></li>
<li class="chapter" data-level="1.11" data-path="inference.html"><a href="inference.html#the-t-distribution-in-practice"><i class="fa fa-check"></i><b>1.11</b> The t-distribution in Practice</a></li>
<li class="chapter" data-level="1.12" data-path="inference.html"><a href="inference.html#confidence-intervals"><i class="fa fa-check"></i><b>1.12</b> Confidence Intervals</a></li>
<li class="chapter" data-level="1.13" data-path="inference.html"><a href="inference.html#power-calculations"><i class="fa fa-check"></i><b>1.13</b> Power Calculations</a></li>
<li class="chapter" data-level="1.14" data-path="inference.html"><a href="inference.html#monte-carlo-simulation"><i class="fa fa-check"></i><b>1.14</b> Monte Carlo Simulation</a></li>
<li class="chapter" data-level="1.15" data-path="inference.html"><a href="inference.html#parametric-simulations-for-the-observations"><i class="fa fa-check"></i><b>1.15</b> Parametric Simulations for the Observations</a></li>
<li class="chapter" data-level="1.16" data-path="inference.html"><a href="inference.html#permutation-tests"><i class="fa fa-check"></i><b>1.16</b> Permutation Tests</a></li>
<li class="chapter" data-level="1.17" data-path="inference.html"><a href="inference.html#association-tests"><i class="fa fa-check"></i><b>1.17</b> Association Tests</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="section-2.html"><a href="section-2.html"><i class="fa fa-check"></i><b>2</b> 探索性数据分析</a><ul>
<li class="chapter" data-level="2.1" data-path="section-2.html"><a href="section-2.html#boxplots"><i class="fa fa-check"></i><b>2.1</b> Boxplots</a></li>
<li class="chapter" data-level="2.2" data-path="section-2.html"><a href="section-2.html#scatterplots-and-correlation"><i class="fa fa-check"></i><b>2.2</b> Scatterplots and Correlation</a></li>
<li class="chapter" data-level="2.3" data-path="section-2.html"><a href="section-2.html#stratification"><i class="fa fa-check"></i><b>2.3</b> Stratification</a></li>
<li class="chapter" data-level="2.4" data-path="section-2.html"><a href="section-2.html#bivariate-normal-distribution"><i class="fa fa-check"></i><b>2.4</b> Bivariate Normal Distribution</a></li>
<li class="chapter" data-level="2.5" data-path="section-2.html"><a href="section-2.html#plots-to-avoid"><i class="fa fa-check"></i><b>2.5</b> Plots to Avoid</a></li>
<li class="chapter" data-level="2.6" data-path="section-2.html"><a href="section-2.html#misunderstanding-correlation-advanced"><i class="fa fa-check"></i><b>2.6</b> Misunderstanding Correlation (Advanced)</a></li>
<li class="chapter" data-level="2.7" data-path="section-2.html"><a href="section-2.html#robust-summaries"><i class="fa fa-check"></i><b>2.7</b> Robust Summaries</a></li>
<li class="chapter" data-level="2.8" data-path="section-2.html"><a href="section-2.html#wilcoxon-rank-sum-test"><i class="fa fa-check"></i><b>2.8</b> Wilcoxon Rank Sum Test</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="matrix-algebra.html"><a href="matrix-algebra.html"><i class="fa fa-check"></i><b>3</b> Matrix Algebra</a><ul>
<li class="chapter" data-level="3.1" data-path="matrix-algebra.html"><a href="matrix-algebra.html#motivating-examples"><i class="fa fa-check"></i><b>3.1</b> Motivating Examples</a></li>
<li class="chapter" data-level="3.2" data-path="matrix-algebra.html"><a href="matrix-algebra.html#matrix-notation"><i class="fa fa-check"></i><b>3.2</b> Matrix Notation</a></li>
<li class="chapter" data-level="3.3" data-path="matrix-algebra.html"><a href="matrix-algebra.html#solving-systems-of-equations"><i class="fa fa-check"></i><b>3.3</b> Solving Systems of Equations</a></li>
<li class="chapter" data-level="3.4" data-path="matrix-algebra.html"><a href="matrix-algebra.html#vectors-matrices-and-scalars"><i class="fa fa-check"></i><b>3.4</b> Vectors, Matrices, and Scalars</a></li>
<li class="chapter" data-level="3.5" data-path="matrix-algebra.html"><a href="matrix-algebra.html#matrix-operations"><i class="fa fa-check"></i><b>3.5</b> Matrix Operations</a></li>
<li class="chapter" data-level="3.6" data-path="matrix-algebra.html"><a href="matrix-algebra.html#examples"><i class="fa fa-check"></i><b>3.6</b> Examples</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="linear-models-1.html"><a href="linear-models-1.html"><i class="fa fa-check"></i><b>4</b> Linear Models</a><ul>
<li class="chapter" data-level="4.1" data-path="linear-models-1.html"><a href="linear-models-1.html#the-design-matrix"><i class="fa fa-check"></i><b>4.1</b> The Design Matrix</a></li>
<li class="chapter" data-level="4.2" data-path="linear-models-1.html"><a href="linear-models-1.html#the-mathematics-behind-lm"><i class="fa fa-check"></i><b>4.2</b> The Mathematics Behind lm()</a></li>
<li class="chapter" data-level="4.3" data-path="linear-models-1.html"><a href="linear-models-1.html#standard-errors"><i class="fa fa-check"></i><b>4.3</b> Standard Errors</a></li>
<li class="chapter" data-level="4.4" data-path="linear-models-1.html"><a href="linear-models-1.html#interactions-and-contrasts"><i class="fa fa-check"></i><b>4.4</b> Interactions and Contrasts</a></li>
<li class="chapter" data-level="4.5" data-path="linear-models-1.html"><a href="linear-models-1.html#linear-model-with-interactions"><i class="fa fa-check"></i><b>4.5</b> Linear Model with Interactions</a></li>
<li class="chapter" data-level="4.6" data-path="linear-models-1.html"><a href="linear-models-1.html#analysis-of-variance"><i class="fa fa-check"></i><b>4.6</b> Analysis of Variance</a></li>
<li class="chapter" data-level="4.7" data-path="linear-models-1.html"><a href="linear-models-1.html#collinearity"><i class="fa fa-check"></i><b>4.7</b> Collinearity</a></li>
<li class="chapter" data-level="4.8" data-path="linear-models-1.html"><a href="linear-models-1.html#rank"><i class="fa fa-check"></i><b>4.8</b> Rank</a></li>
<li class="chapter" data-level="4.9" data-path="linear-models-1.html"><a href="linear-models-1.html#removing-confounding"><i class="fa fa-check"></i><b>4.9</b> Removing Confounding</a></li>
<li class="chapter" data-level="4.10" data-path="linear-models-1.html"><a href="linear-models-1.html#the-qr-factorization-advanced"><i class="fa fa-check"></i><b>4.10</b> The QR Factorization (Advanced)</a></li>
<li class="chapter" data-level="4.11" data-path="linear-models-1.html"><a href="linear-models-1.html#going-further"><i class="fa fa-check"></i><b>4.11</b> Going Further</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="inference-for-high-dimensional-data.html"><a href="inference-for-high-dimensional-data.html"><i class="fa fa-check"></i><b>5</b> Inference for High Dimensional Data</a><ul>
<li class="chapter" data-level="5.1" data-path="inference-for-high-dimensional-data.html"><a href="inference-for-high-dimensional-data.html#introduction-3"><i class="fa fa-check"></i><b>5.1</b> Introduction</a></li>
<li class="chapter" data-level="5.2" data-path="inference-for-high-dimensional-data.html"><a href="inference-for-high-dimensional-data.html#inference-in-practice"><i class="fa fa-check"></i><b>5.2</b> Inference in Practice</a></li>
<li class="chapter" data-level="5.3" data-path="inference-for-high-dimensional-data.html"><a href="inference-for-high-dimensional-data.html#procedures"><i class="fa fa-check"></i><b>5.3</b> Procedures</a></li>
<li class="chapter" data-level="5.4" data-path="inference-for-high-dimensional-data.html"><a href="inference-for-high-dimensional-data.html#error-rates"><i class="fa fa-check"></i><b>5.4</b> Error Rates</a></li>
<li class="chapter" data-level="5.5" data-path="inference-for-high-dimensional-data.html"><a href="inference-for-high-dimensional-data.html#the-bonferroni-correction"><i class="fa fa-check"></i><b>5.5</b> The Bonferroni Correction</a></li>
<li class="chapter" data-level="5.6" data-path="inference-for-high-dimensional-data.html"><a href="inference-for-high-dimensional-data.html#false-discovery-rate"><i class="fa fa-check"></i><b>5.6</b> False Discovery Rate</a></li>
<li class="chapter" data-level="5.7" data-path="inference-for-high-dimensional-data.html"><a href="inference-for-high-dimensional-data.html#direct-approach-to-fdr-and-q-values-advanced"><i class="fa fa-check"></i><b>5.7</b> Direct Approach to FDR and q-values (Advanced)</a></li>
<li class="chapter" data-level="5.8" data-path="inference-for-high-dimensional-data.html"><a href="inference-for-high-dimensional-data.html#basic-exploratory-data-analysis"><i class="fa fa-check"></i><b>5.8</b> Basic Exploratory Data Analysis</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="section-6.html"><a href="section-6.html"><i class="fa fa-check"></i><b>6</b> 统计模型</a><ul>
<li class="chapter" data-level="6.1" data-path="section-6.html"><a href="section-6.html#-the-binomial-distribution"><i class="fa fa-check"></i><b>6.1</b> 二项分布 (The Binomial Distribution)</a></li>
<li class="chapter" data-level="6.2" data-path="section-6.html"><a href="section-6.html#-the-poisson-distribution"><i class="fa fa-check"></i><b>6.2</b> 泊松分布 (The Poisson Distribution)</a></li>
<li class="chapter" data-level="6.3" data-path="section-6.html"><a href="section-6.html#section-6.3"><i class="fa fa-check"></i><b>6.3</b> 最大似然估计</a></li>
<li class="chapter" data-level="6.4" data-path="section-6.html"><a href="section-6.html#section-6.4"><i class="fa fa-check"></i><b>6.4</b> 连续变量的分布</a></li>
<li class="chapter" data-level="6.5" data-path="section-6.html"><a href="section-6.html#section-6.5"><i class="fa fa-check"></i><b>6.5</b> 贝叶斯统计</a></li>
<li class="chapter" data-level="6.6" data-path="section-6.html"><a href="section-6.html#hierarchical-models"><i class="fa fa-check"></i><b>6.6</b> Hierarchical Models</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="section-7.html"><a href="section-7.html"><i class="fa fa-check"></i><b>7</b> 距离和维度降低   </a><ul>
<li class="chapter" data-level="7.1" data-path="section-7.html"><a href="section-7.html#-1"><i class="fa fa-check"></i><b>7.1</b> 简介   </a></li>
<li class="chapter" data-level="7.2" data-path="section-7.html"><a href="section-7.html#euclidean-distance"><i class="fa fa-check"></i><b>7.2</b> Euclidean Distance</a></li>
<li class="chapter" data-level="7.3" data-path="section-7.html"><a href="section-7.html#section-7.3"><i class="fa fa-check"></i><b>7.3</b> 高维数据的距离   </a></li>
<li class="chapter" data-level="7.4" data-path="section-7.html"><a href="section-7.html#distance-exercises"><i class="fa fa-check"></i><b>7.4</b> Distance exercises</a></li>
<li class="chapter" data-level="7.5" data-path="section-7.html"><a href="section-7.html#dimension-reduction-motivation"><i class="fa fa-check"></i><b>7.5</b> Dimension Reduction Motivation</a></li>
<li class="chapter" data-level="7.6" data-path="section-7.html"><a href="section-7.html#singular-value-decomposition"><i class="fa fa-check"></i><b>7.6</b> Singular Value Decomposition</a></li>
<li class="chapter" data-level="7.7" data-path="section-7.html"><a href="section-7.html#projections"><i class="fa fa-check"></i><b>7.7</b> Projections</a></li>
<li class="chapter" data-level="7.8" data-path="section-7.html"><a href="section-7.html#rotations-1"><i class="fa fa-check"></i><b>7.8</b> Rotations</a></li>
<li class="chapter" data-level="7.9" data-path="section-7.html"><a href="section-7.html#multi-dimensional-scaling-plots"><i class="fa fa-check"></i><b>7.9</b> Multi-Dimensional Scaling Plots</a></li>
<li class="chapter" data-level="7.10" data-path="section-7.html"><a href="section-7.html#mds-exercises"><i class="fa fa-check"></i><b>7.10</b> MDS exercises</a></li>
<li class="chapter" data-level="7.11" data-path="section-7.html"><a href="section-7.html#principal-component-analysis"><i class="fa fa-check"></i><b>7.11</b> Principal Component Analysis</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="basic-machine-learning.html"><a href="basic-machine-learning.html"><i class="fa fa-check"></i><b>8</b> Basic Machine Learning</a><ul>
<li class="chapter" data-level="8.1" data-path="basic-machine-learning.html"><a href="basic-machine-learning.html#clustering"><i class="fa fa-check"></i><b>8.1</b> Clustering</a></li>
<li class="chapter" data-level="8.2" data-path="basic-machine-learning.html"><a href="basic-machine-learning.html#conditional-probabilities-and-expectations"><i class="fa fa-check"></i><b>8.2</b> Conditional Probabilities and Expectations</a></li>
<li class="chapter" data-level="8.3" data-path="basic-machine-learning.html"><a href="basic-machine-learning.html#smoothing"><i class="fa fa-check"></i><b>8.3</b> Smoothing</a></li>
<li class="chapter" data-level="8.4" data-path="basic-machine-learning.html"><a href="basic-machine-learning.html#bin-smoothing"><i class="fa fa-check"></i><b>8.4</b> Bin Smoothing</a></li>
<li class="chapter" data-level="8.5" data-path="basic-machine-learning.html"><a href="basic-machine-learning.html#loess"><i class="fa fa-check"></i><b>8.5</b> Loess</a></li>
<li class="chapter" data-level="8.6" data-path="basic-machine-learning.html"><a href="basic-machine-learning.html#class-prediction"><i class="fa fa-check"></i><b>8.6</b> Class Prediction</a></li>
<li class="chapter" data-level="8.7" data-path="basic-machine-learning.html"><a href="basic-machine-learning.html#cross-validation"><i class="fa fa-check"></i><b>8.7</b> Cross-validation</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="batch-effects.html"><a href="batch-effects.html"><i class="fa fa-check"></i><b>9</b> Batch Effects</a><ul>
<li class="chapter" data-level="9.1" data-path="batch-effects.html"><a href="batch-effects.html#confounding"><i class="fa fa-check"></i><b>9.1</b> Confounding</a></li>
<li class="chapter" data-level="9.2" data-path="batch-effects.html"><a href="batch-effects.html#confounding-high-throughput-example"><i class="fa fa-check"></i><b>9.2</b> Confounding: High-Throughput Example</a></li>
<li class="chapter" data-level="9.3" data-path="batch-effects.html"><a href="batch-effects.html#discovering-batch-effects-with-eda"><i class="fa fa-check"></i><b>9.3</b> Discovering Batch Effects with EDA</a></li>
<li class="chapter" data-level="9.4" data-path="batch-effects.html"><a href="batch-effects.html#gene-expression-data"><i class="fa fa-check"></i><b>9.4</b> Gene Expression Data</a></li>
<li class="chapter" data-level="9.5" data-path="batch-effects.html"><a href="batch-effects.html#motivation-for-statistical-approaches"><i class="fa fa-check"></i><b>9.5</b> Motivation for Statistical Approaches</a></li>
<li class="chapter" data-level="9.6" data-path="batch-effects.html"><a href="batch-effects.html#adjusting-for-batch-effects-with-linear-models"><i class="fa fa-check"></i><b>9.6</b> Adjusting for Batch Effects with Linear Models</a></li>
<li class="chapter" data-level="9.7" data-path="batch-effects.html"><a href="batch-effects.html#factor-analysis"><i class="fa fa-check"></i><b>9.7</b> Factor Analysis</a></li>
<li class="chapter" data-level="9.8" data-path="batch-effects.html"><a href="batch-effects.html#modeling-batch-effects-with-factor-analysis"><i class="fa fa-check"></i><b>9.8</b> Modeling Batch Effects with Factor Analysis</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>参考文献</a></li>
<li class="divider"></li>
<li><a href="https://bookdown.org" target="blank">本书由 bookdown 强力驱动</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">生物信息R数据分析</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="acknowledgments" class="section level1 unnumbered">
<h1>Acknowledgments</h1>
<p>The authors would like to thank Alex Nones for proofreading the manuscript during its various stages. Also, thanks to Karl Broman for contributing the “Plots to Avoid” section and to Stephanie Hicks for designing some of the exercises. Finally, thanks to John Kimmel and three anonymous referees for excellent feedback and constructive criticism of the book.</p>
<p>This book was conceived during the teaching of several HarvardX courses, coordinated by Heather Sternshein. We are also grateful to our TAs, Idan Ginsburg and Stephanie Chan, and all the students whose questions and comments helped us improve the book. The courses were partially funded by NIH grant R25GM114818. We are very grateful to the National Institute of Health for its support.</p>
<p>A special thanks goes to all those that edited the book via GitHub pull requests: vjcitn, yeredh, ste-fan, molx, kern3020, josemrecio, hcorrada, neerajt, massie, jmgore75, molecules, lzamparo, eronisko, obicke, knbknb, and devrajoh.</p>
<p>Cover image credit: this photograph is La Mina Falls, El Yunque National Forest, Puerto Rico, taken by Ron Kroetz <a href="https://www.flickr.com/photos/ronkroetz/14779273923" class="uri">https://www.flickr.com/photos/ronkroetz/14779273923</a> Attribution-NoDerivs 2.0 Generic (CC BY-ND 2.0)</p>
<p>{mainmatter}</p>
<div id="section-0.1" class="section level2">
<h2><span class="header-section-number">0.1</span> 简介</h2>
<p>  在二十世纪的下半叶，数字技术的革新带来了测量技术的革新，从而也大大的改变了科研领域。在生命科学领域，数据分析已经是科研项目中的家常便饭了。尤其是基因组学，新的技术使得我们们能够观察到我们过去没有办法观察到的特定的。历史上显微镜的发明使得我们能够鉴定新的微生物，而近些年的新技术技术引领我们发现更多生命科学的奥秘，而这些发现不亚于之前显微镜发明带给我们的帮助。这些技术中的代表就是芯片技术和下一代测序技术。</p>
<p>  这些技术使得一些传统的依赖简单的数据分析的科研领域发生了翻天覆地的变化。例如，过去科学家如果对某一个基因感兴趣。，可能会测量其表达量。今天，利用新的技术我们可以一次性测量超过2万个人类基因的表达量。以前我们的研究是受假设驱动的，技术进步使得科研领域从</p>
<p>  由于越来越多的研究开始依赖于上面我们提到的数据，许多生命科学领域的科研工作者开始转变成数据分析专家。这本书就是由这样的一批科研工作者写的。如果你自己分析自己的数据，那么你很可能会计算P值，应用Bonferroni矫正，进行主成分分析，产生一个热图等等。如果你发现你并不是十分理解这些技术，或者你不确定你应用这些技术时是否恰当，那么这本书就是为你写的。</p>
<p>  尽管这本书大部分是着重于高级的统计概念，但是我们是从最基本的知识开始的，从而确保所有的读者能够有一个扎实的数据分析的统计基础。我们发现许多基础统计课程的方式是读者很难讲概念性的东西和真正的数据分析联系起来。我们的方法会确保你能学习到实践和理论的结合。所以，前两章的内容（推断与探索性数据分析）适用于本科阶段的统计或者数据科学的入门课程。接着，统计的难度就会慢慢的增加。</p>
<p>  尽管本书的读者通常会有硕士或者博士学位，我们会尽量将数学的内容保持在本科入门水平。这本书中你不需要用到微积分。然而，我们会介绍和使用线性代数，而线性代数通常被认为比微积分难的一门学科。但是我们会在数据分析的背景下来理解线性代数，我们相信在这种情况下，你可以在不懂微积分的情况下理解基础的线性代数知识。比较难的部分可能是你需要适应符号和公式。——–</p>
</div>
<div id="e8bf99e69cace4b9a6e58c85e590abe4bb80e4b988e58685e5aeb9efbc9f" class="section level2 unnumbered">
<h2>这本书包含什么内容？</h2>
<p>&amp;emsp&amp;emsp这本书包含以数据为主的生命科学研究中的一些统计概念和数据分析的技巧。我们从简单的P值计算开始，一步步到分析高通量数据的高级内容。</p>
<p>&amp;emsp&amp;emsp我们首先会介绍统计和生命科学领域中的一个最重要的内容：统计推断。统计推断是利用概率来掌握数据的群体特征。一个经典的例子是判断两组数据的均值是否有差异。其他的内容包括t检验，置信区间，联想测验（association test），蒙特卡洛方法，置换检验和统计效力。我们会使用经典数学理论带给我们的中心极限定理和当代计算发展……………</p>
<p>We then cover the concepts of distance and dimension reduction. We will introduce the mathematical definition of distance and use this to motivate the singular value decomposition (SVD) for dimension reduction and multi-dimensional scaling. Once we learn this, we will be ready to cover hierarchical and k-means clustering. We will follow this with a basic introduction to machine learning.</p>
<p>We end by learning about batch effects and how component and factor analysis are used to deal with this challenge. In particular, we will examine confounding, show examples of batch effects, make the connection to factor analysis, and describe surrogate variable analysis.</p>
</div>
<div id="how-is-this-book-different" class="section level2 unnumbered">
<h2>How Is This Book Different?</h2>
</div>
<div id="section-0.2" class="section level2">
<h2><span class="header-section-number">0.2</span> 这本书有什么不同？</h2>
<p>While statistics textbooks focus on mathematics, this book focuses on using a computer to perform data analysis. This book follows the approach of <a href="https://www.stat.berkeley.edu/~statlabs/">Stat Labs</a>, by Deborah Nolan and Terry Speed. Instead of explaining the mathematics and theory, and then showing examples, we start by stating a practical data-related challenge. This book also includes the computer code that provides a solution to the problem and helps illustrate the concepts behind the solution. By running the code yourself, and seeing data generation and analysis happen live, you will get a better intuition for the concepts, the mathematics, and the theory.</p>
<p>通常统计相关的课本重点放在数学公式上，这本书重点是利用计算机来进行数据分析。这本书所遵循的方法与Deborah Nolan和Terry Speed所著的<a href="https://www.stat.berkeley.edu/~statlabs/">Stat Labs</a>是一样的。通常统计课本是先解释数学和理论知识，然后展示例子，我们是从一个具体的实践中会遇到的数据分析问题开始。这本书也包含解决这些问题的具体的R代码，这些代码可以帮助读者理解解决方法。</p>
<p>We focus on the practical challenges faced by data analysts in the life sciences and introduce mathematics as a tool that can help us achieve scientific goals. Furthermore, throughout the book we show the R code that performs this analysis and connect the lines of code to the statistical and mathematical concepts we explain. All sections of this book are reproducible as they were made using <em>R markdown</em> documents that include R code used to produce the figures, tables and results shown in the book. In order to distinguish it, the code is shown in the following font:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="dv">2</span> 
y &lt;-<span class="st"> </span><span class="dv">3</span> 
<span class="kw">print</span>(x<span class="op">+</span>y) </code></pre></div>
<p>and the results in different colors, preceded by two hash characters (<em>##</em>):</p>
<pre><code>## [1] 5</code></pre>
<p>We will provide links that will give you access to the raw R markdown code so you can easily follow along with the book by programming in R.</p>
<p>At the beginning of each chapter you will see the sentence:</p>
<blockquote>
<blockquote>
<p>The R markdown document for this section is available here.</p>
</blockquote>
</blockquote>
<p>The word “here” will be a hyperlink to the R markdown file. The best way to read this book is with a computer in front of you, scrolling through that file, and running the R code that produces the results included in the book section you are reading.</p>

<p>  本书中所有的数据分析我们都使用<a href="https://cran.r-project.org/">R语言</a>。我们假设你有基础的编程能力和R语言语法的知识。如果你没有这些基础知识，你第一个要做的就是学习R的基础知识。这里我们会从安装R开始一步一步的教你如何使用R。</p>
</div>
<div id="r" class="section level2">
<h2><span class="header-section-number">0.3</span> 安装R</h2>
<p>  第一步是安装R。你可以从R的官网上下载R。安装的过程非常简单，但是如果你需要更多的帮助，你可以下载下面对应的资源。</p>
<ul>
<li><a href="https://github.com/genomicsclass/windows#installing-r">Windows系统R安装教程</a></li>
<li><a href="http://youtu.be/Icawuhf0Yqo">Mac系统R安装教程</a></li>
<li><a href="http://cran.r-project.org/bin/linux/ubuntu/README">Ubuntu系统R安装教程</a></li>
</ul>
<p>  下面的一步是安装Rstudio。Rstudio。在不安装Rstudio的情况下，你可以运行本书中所有的R代码，但是我们强烈推荐使用Rstudio这款集成开发环境软件。具体的下载链接见<a href="http://www.rstudio.com/products/rstudio/download/">这里</a>(<a href="http://www.rstudio.com/products/rstudio/download/" class="uri">http://www.rstudio.com/products/rstudio/download/</a>)。对于Windows系统，我们也有一个对应的<a href="https://github.com/genomicsclass/windows">安装手册</a>：<a href="https://github.com/genomicsclass/windows" class="uri">https://github.com/genomicsclass/windows</a>.</p>
</div>
<div id="r" class="section level2">
<h2><span class="header-section-number">0.4</span> R基础知识</h2>
<p>  你首先要做的是让你自己对R编程和R的语法熟悉起来，你可以通过完成一个R的基本教程来实现这个目的。为了能够理解本书，你需要理解列表（list），数值向量（numeric vectors），能够写函数，以及理解如何使用<code>sapply</code>和<code>replicate</code>这两个函数。</p>
<p>  如果你已经对R非常熟悉，你可以跳过下一节。如果你不熟悉，我们建议你完成<a href="http://swirlstats.com/">swirl</a>教程，这个教程会以交互的方式教你R编程和数字科学。在你安装R后，你可以安装<code>swirl</code>，然后以下面的方式运行它：</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">install.packages</span>(<span class="st">&quot;swirl&quot;</span>)
<span class="kw">library</span>(swirl)
<span class="kw">swirl</span>()</code></pre></div>
<p>  你也可以尝试Code School的<a href="http://tryr.codeschool.com/">try R</a>课程。</p>
<p>&amp;emsp&amp;emsp同时，你也可以找到许多学习R的开源的资源和手册。这里我们给出两个例子：</p>
<ul>
<li><a href="http://www.statmethods.net/">R快速入门（Quick-R）</a>: a quick online reference for data input, basic statistics and plots</li>
<li>R参考卡（R reference card） (PDF)[<a href="https://cran.r-project.org/doc/contrib/Short-refcard.pdf" class="uri">https://cran.r-project.org/doc/contrib/Short-refcard.pdf</a>] by Tom Short</li>
</ul>
<p>&amp;emsp&amp;emsp关于R有两个关键的内容：一是你能够使用 <code>help``?</code></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">?install.packages
<span class="kw">help</span>(<span class="st">&quot;install.packages&quot;</span>)</code></pre></div>
<p>&amp;emsp&amp;emsp另一个就是#号表示注释，#号后面的内容是不会执行的。注释对编程代码的可读性是非常重要的。良好的注释是优秀代码的主要组成部分。</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">##This is just a comment</code></pre></div>
</div>
<div id="installing-packages" class="section level2">
<h2><span class="header-section-number">0.5</span> Installing Packages</h2>
<p>The first R command we will run is <code>install.packages</code>. If you took the <code>swirl</code> tutorial you should have already done this. R only includes a basic set of functions. It can do much more than this, but not everybody needs everything so we instead make some functions available via packages. Many of these functions are stored in CRAN. Note that these packages are vetted: they are checked for common errors and they must have a dedicated maintainer. You can easily install packages from within R if you know the name of the packages. As an example, we are going to install the package <code>rafalib</code> which we use in our first data analysis examples:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">install.packages</span>(<span class="st">&quot;rafalib&quot;</span>)</code></pre></div>
<p>We can then load the package into our R sessions using the <code>library</code> function:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(rafalib)</code></pre></div>
<p>From now on you will see that we sometimes load packages without installing them. This is because once you install the package, it remains in place and only needs to be loaded with <code>library</code>. If you try to load a package and get an error, it probably means you need to install it first.</p>
</div>
<div id="importing-data-into-r" class="section level2">
<h2><span class="header-section-number">0.6</span> Importing Data into R</h2>
<p>The first step when preparing to analyze data is to read in the data into R. There are several ways to do this and we will discuss three of them. But you only need to learn one to follow along.</p>
<p>In the life sciences, small datasets such as the one used as an example in the next sections are typically stored as Excel files. Although there are R packages designed to read Excel (xls) format, you generally want to avoid this and save files as comma delimited (Comma-Separated Value/CSV) or tab delimited (Tab-Separated Value/TSV/TXT) files. These plain-text formats are often easier for sharing data with collaborators, as commercial software is not required for viewing or working with the data. We will start with a simple example dataset containing <a href="https://raw.githubusercontent.com/genomicsclass/dagdata/master/inst/extdata/femaleMiceWeights.csv">female mouse weights</a>.</p>
<p>第一步是找到对应的文件并且弄清楚这个文件的<em>路径</em>。</p>
<div id="section-0.6.0.1" class="section level4">
<h4><span class="header-section-number">0.6.0.1</span> 路径和工作目录</h4>
<p>当你在</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">getwd</span>()</code></pre></div>
<p>You can also change your working directory using the function <code>setwd</code>. Or you can change it through RStudio by clicking on “Session”.</p>
<p>The functions that read and write files (there are several in R) assume you mean to look for files or write files in the working directory. Our recommended approach for beginners will have you reading and writing to the working directory. However, you can also type the <a href="http://www.computerhope.com/jargon/a/absopath.htm">full path</a>, which will work independently of the working directory.</p>
</div>
<div id="projects-in-rstudio" class="section level4">
<h4><span class="header-section-number">0.6.0.2</span> Projects in RStudio</h4>
<p>We find that the simplest way to organize yourself is to start a Project in RStudio (Click on “File” and “New Project”). When creating the project, you will select a folder to be associated with it. You can then download all your data into this folder. Your working directory will be this folder.</p>
</div>
<div id="option-1-read-file-over-the-internet" class="section level4">
<h4><span class="header-section-number">0.6.0.3</span> Option 1: Read file over the Internet</h4>
<p>You can navigate to the <code>femaleMiceWeights.csv</code> file by visiting the data directory of dagdata on <a href="https://github.com/genomicsclass/dagdata/tree/master/inst/extdata">GitHub</a>. If you navigate to the file, you need to click on <em>Raw</em> on the upper right hand corner of the page.</p>
<div class="figure">
<img src="images/handmade/screenshot1.png" alt="GitHub page screenshot." />
<p class="caption">GitHub page screenshot.</p>
</div>
<p>Now you can copy and paste the URL and use this as the argument to <code>read.csv</code>. Here we break the URL into a base directory and a filename and then combine with <code>paste0</code> because the URL would otherwise be too long for the page. We use <code>paste0</code> because we want to put the strings together as is, if you were specifying a file on your machine you should use the smarter function, <code>file.path</code>, which knows the difference between Windows and Mac file path connectors. You can specify the URL using a single string to avoid this extra step.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">dir &lt;-<span class="st"> &quot;https://raw.githubusercontent.com/genomicsclass/dagdata/master/inst/extdata/&quot;</span>
url &lt;-<span class="st"> </span><span class="kw">paste0</span>(dir, <span class="st">&quot;femaleMiceWeights.csv&quot;</span>)
dat &lt;-<span class="st"> </span><span class="kw">read.csv</span>(url)</code></pre></div>
</div>
<div id="option-2-download-file-with-your-browser-to-your-working-directory" class="section level4">
<h4><span class="header-section-number">0.6.0.4</span> Option 2: Download file with your browser to your working directory</h4>
<p>There are reasons for wanting to keep a local copy of the file. For example, you may want to run the analysis while not connected to the Internet or you may want to ensure reproducibility regardless of the file being available on the original site. To download the file, as in option 1, you can navigate to the <code>femaleMiceWeights.csv</code>. In this option we use your browser’s “Save As” function to ensure that the downloaded file is in a CSV format. Some browsers add an extra suffix to your filename by default. You do not want this. You want your file to be named <code>femaleMiceWeights.csv</code>. Once you have this file in your working directory, then you can simply read it in like this:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">dat &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="st">&quot;femaleMiceWeights.csv&quot;</span>)</code></pre></div>
<p>If you did not receive any message, then you probably read in the file successfully.</p>
</div>
<div id="option-3-download-the-file-from-within-r" class="section level4">
<h4><span class="header-section-number">0.6.0.5</span> Option 3: Download the file from within R</h4>
<p>We store many of the datasets used here on <a href="https://github.com/genomicsclass/">GitHub</a>. You can save these files directly from the Internet to your computer using R. In this example, we are using the <code>download.file</code> function in the <code>downloader</code> package to download the file to a specific location and then read it in. We can assign it a random name and a random directory using the function <code>tempfile</code>, but you can also save it in directory with the name of your choosing.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(downloader) ##use install.packages to install
dir &lt;-<span class="st"> &quot;https://raw.githubusercontent.com/genomicsclass/dagdata/master/inst/extdata/&quot;</span>
filename &lt;-<span class="st"> &quot;femaleMiceWeights.csv&quot;</span> 
url &lt;-<span class="st"> </span><span class="kw">paste0</span>(dir, filename)
<span class="cf">if</span> (<span class="op">!</span><span class="kw">file.exists</span>(filename)) <span class="kw">download</span>(url, <span class="dt">destfile=</span>filename)</code></pre></div>
<p>We can then proceed as in option 2:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">dat &lt;-<span class="st"> </span><span class="kw">read.csv</span>(filename)</code></pre></div>
</div>
<div id="option-4-download-the-data-package-advanced" class="section level4">
<h4><span class="header-section-number">0.6.0.6</span> Option 4: Download the data package (Advanced)</h4>
<p>Many of the datasets we include in this book are available in custom-built packages from GitHub. The reason we use GitHub, rather than CRAN, is that on GitHub we do not have to vet packages, which gives us much more flexibility.</p>
<p>To install packages from GitHub you will need to install the <code>devtools</code> package:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">install.packages</span>(<span class="st">&quot;devtools&quot;</span>)</code></pre></div>
<p>Note to Windows users: to use devtools you will have to also install <code>Rtools</code>. In general you will need to install packages as administrator. One way to do this is to start R as administrator. If you do not have permission to do this, then it is a bit <a href="http://www.magesblog.com/2012/04/installing-r-packages-without-admin.html">more complicated</a>.</p>
<p>Now you are ready to install a package from GitHub. For this we use a different function:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(devtools)
<span class="kw">install_github</span>(<span class="st">&quot;genomicsclass/dagdata&quot;</span>)</code></pre></div>
<p>The file we are working with is actually included in this package. Once you install the package, the file is on your computer. However, finding it requires advanced knowledge. Here are the lines of code:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">dir &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="dt">package=</span><span class="st">&quot;dagdata&quot;</span>) <span class="co">#extracts the location of package</span>
<span class="kw">list.files</span>(dir)</code></pre></div>
<pre><code>## [1] &quot;data&quot;        &quot;DESCRIPTION&quot; &quot;extdata&quot;    
## [4] &quot;help&quot;        &quot;html&quot;        &quot;Meta&quot;       
## [7] &quot;NAMESPACE&quot;   &quot;script&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">list.files</span>(<span class="kw">file.path</span>(dir,<span class="st">&quot;extdata&quot;</span>)) <span class="co">#external data is in this directory</span></code></pre></div>
<pre><code>## [1] &quot;admissions.csv&quot;              
## [2] &quot;astronomicalunit.csv&quot;        
## [3] &quot;babies.txt&quot;                  
## [4] &quot;femaleControlsPopulation.csv&quot;
## [5] &quot;femaleMiceWeights.csv&quot;       
## [6] &quot;mice_pheno.csv&quot;              
## [7] &quot;msleep_ggplot2.csv&quot;          
## [8] &quot;README&quot;                      
## [9] &quot;spider_wolff_gorb_2013.csv&quot;</code></pre>
<p>And now we are ready to read in the file:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">filename &lt;-<span class="st"> </span><span class="kw">file.path</span>(dir,<span class="st">&quot;extdata/femaleMiceWeights.csv&quot;</span>)
dat &lt;-<span class="st"> </span><span class="kw">read.csv</span>(filename)</code></pre></div>
</div>
<div id="getting-started-exercises" class="section level3">
<h3><span class="header-section-number">0.6.1</span> Getting Started Exercises</h3>
<p><strong>Exercises</strong></p>
<p>Here we will test some of the basics of R data manipulation which you should know or should have learned by following the tutorials above. You will need to have the file <code>femaleMiceWeights.csv</code> in your working directory. As we showed above, one way to do this is by using the <code>downloader</code> package:</p>
<pre><code>library(downloader) 
url &lt;- &quot;https://raw.githubusercontent.com/genomicsclass/dagdata/master/inst/extdata/femaleMiceWeights.csv&quot;
filename &lt;- &quot;femaleMiceWeights.csv&quot; 
download(url, destfile=filename)</code></pre>
<ol style="list-style-type: decimal">
<li><p>Read in the <code>file femaleMiceWeights.csv</code> and report the body weight of the mouse in the exact name of the column containing the weights.</p></li>
<li><p>The <code>[</code> and <code>]</code> symbols can be used to extract specific rows and specific columns of the table. What is the entry in the 12th row and second column?</p></li>
<li><p>You should have learned how to use the <code>$</code> character to extract a column from a table and return it as a vector. Use <code>$</code> to extract the weight column and report the weight of the mouse in the 11th row.</p></li>
<li><p>The length function returns the number of elements in a vector. How many mice are included in our dataset?</p></li>
<li><p>To create a vector with the numbers 3 to 7, we can use seq(3,7) or, because they are consecutive, 3:7. View the data and determine what rows are associated with the high fat or hf diet. Then use the mean function to compute the average weight of these mice.</p></li>
<li><p>One of the functions we will be using often is sample. Read the help file for sample using ?sample. Now take a random sample of size 1 from the numbers 13 to 24 and report back the weight of the mouse represented by that row. Make sure to type set.seed(1) to ensure that everybody gets the same answer.</p></li>
</ol>
</div>
</div>
<div id="brief-introduction-to-dplyr" class="section level2">
<h2><span class="header-section-number">0.7</span> Brief Introduction to <code>dplyr</code></h2>
<p>The learning curve for R syntax is slow. One of the more difficult aspects that requires some getting used to is subsetting data tables. The <code>dplyr</code> package brings these tasks closer to English and we are therefore going to introduce two simple functions: one is used to subset and the other to select columns.</p>
<p>Take a look at the dataset we read in:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">filename &lt;-<span class="st"> &quot;femaleMiceWeights.csv&quot;</span>
dat &lt;-<span class="st"> </span><span class="kw">read.csv</span>(filename)
<span class="kw">head</span>(dat) <span class="co">#In R Studio use View(dat)</span></code></pre></div>
<pre><code>##   Diet Bodyweight
## 1 chow      21.51
## 2 chow      28.14
## 3 chow      24.04
## 4 chow      23.45
## 5 chow      23.68
## 6 chow      19.79</code></pre>
<p>There are two types of diets, which are denoted in the first column. If we want just the weights, we only need the second column. So if we want the weights for mice on the <code>chow</code> diet, we subset and filter like this:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(dplyr) 
chow &lt;-<span class="st"> </span><span class="kw">filter</span>(dat, Diet<span class="op">==</span><span class="st">&quot;chow&quot;</span>) <span class="co">#keep only the ones with chow diet</span>
<span class="kw">head</span>(chow)</code></pre></div>
<pre><code>##   Diet Bodyweight
## 1 chow      21.51
## 2 chow      28.14
## 3 chow      24.04
## 4 chow      23.45
## 5 chow      23.68
## 6 chow      19.79</code></pre>
<p>And now we can select only the column with the values:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">chowVals &lt;-<span class="st"> </span><span class="kw">select</span>(chow,Bodyweight)
<span class="kw">head</span>(chowVals)</code></pre></div>
<pre><code>##   Bodyweight
## 1      21.51
## 2      28.14
## 3      24.04
## 4      23.45
## 5      23.68
## 6      19.79</code></pre>
<p>A nice feature of the <code>dplyr</code> package is that you can perform consecutive tasks by using what is called a “pipe”. In <code>dplyr</code> we use <code>%&gt;%</code> to denote a pipe. This symbol tells the program to first do one thing and then do something else to the result of the first. Hence, we can perform several data manipulations in one line. For example:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">chowVals &lt;-<span class="st"> </span><span class="kw">filter</span>(dat, Diet<span class="op">==</span><span class="st">&quot;chow&quot;</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">select</span>(Bodyweight)</code></pre></div>
<p>In the second task, we no longer have to specify the object we are editing since it is whatever comes from the previous call.</p>
<p>Also, note that if <code>dplyr</code> receives a <code>data.frame</code> it will return a <code>data.frame</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">class</span>(dat)</code></pre></div>
<pre><code>## [1] &quot;data.frame&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">class</span>(chowVals)</code></pre></div>
<pre><code>## [1] &quot;data.frame&quot;</code></pre>
<p>For pedagogical reasons, we will often want the final result to be a simple <code>numeric</code> vector. To obtain such a vector with <code>dplyr</code>, we can apply the <code>unlist</code> function which turns <code>lists</code>, such as <code>data.frames</code>, into <code>numeric</code> vectors:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">chowVals &lt;-<span class="st"> </span><span class="kw">filter</span>(dat, Diet<span class="op">==</span><span class="st">&quot;chow&quot;</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">select</span>(Bodyweight) <span class="op">%&gt;%</span><span class="st"> </span>unlist
<span class="kw">class</span>( chowVals )</code></pre></div>
<pre><code>## [1] &quot;numeric&quot;</code></pre>
<p>To do this in R without <code>dplyr</code> the code is the following:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">chowVals &lt;-<span class="st"> </span>dat[ dat<span class="op">$</span>Diet<span class="op">==</span><span class="st">&quot;chow&quot;</span>, <span class="kw">colnames</span>(dat)<span class="op">==</span><span class="st">&quot;Bodyweight&quot;</span>]</code></pre></div>
<div id="dplyr-exercises" class="section level3">
<h3><span class="header-section-number">0.7.1</span> <code>dplyr</code> exercises</h3>
<p><strong>Exercises</strong></p>
<p>For these exercises, we will use a new dataset related to mammalian sleep. This data is described <a href="http://ggplot2.tidyverse.org/reference/msleep.html">here</a>. Download the CSV file from this location:</p>
<p>We are going to read in this data, then test your knowledge of they key <code>dplyr</code> functions <code>select</code> and <code>filter</code>. We are also going to review two different classes: data frames and vectors.</p>
<ol style="list-style-type: decimal">
<li><p>Read in the <code>msleep_ggplot2.csv</code> file with the function <code>read.csv</code> and use the function <code>class</code> to determine what type of object is returned.</p></li>
<li><p>Now use the <code>filter</code> function to select only the primates. How many animals in the table are primates? Hint: the nrow function gives you the number of rows of a data frame or matrix.</p></li>
<li><p>What is the class of the object you obtain after subsetting the table to only include primates?</p></li>
<li><p>Now use the <code>select</code> function to extract the sleep (total) for the primates. What class is this object? Hint: use <code>%&gt;%</code> to pipe the results of the <code>filter</code> function to <code>select</code>.</p></li>
<li><p>Now we want to calculate the average amount of sleep for primates (the average of the numbers computed above). One challenge is that the <code>mean</code> function requires a vector so, if we simply apply it to the output above, we get an error. Look at the help file for <code>unlist</code> and use it to compute the desired average.</p></li>
<li><p>For the last exercise, we could also use the dplyr <code>summarize</code> function. We have not introduced this function, but you can read the help file and repeat exercise 5, this time using just <code>filter</code> and <code>summarize</code> to get the answer.</p></li>
</ol>
</div>
</div>
<div id="mathematical-notation" class="section level2">
<h2><span class="header-section-number">0.8</span> Mathematical Notation</h2>
<p>This book focuses on teaching statistical concepts and data analysis programming skills. We avoid mathematical notation as much as possible, but we do use it. We do not want readers to be intimidated by the notation though. Mathematics is actually the easier part of learning statistics. Unfortunately, many text books use mathematical notation in what we believe to be an over-complicated way. For this reason, we do try to keep the notation as simple as possible. However, we do not want to water down the material, and some mathematical notation facilitates a deeper understanding of the concepts. Here we describe a few specific symbols that we use often. If they appear intimidating to you, please take some time to read this section carefully as they are actually simpler than they seem. Because by now you should be somewhat familiar with R, we will make the connection between mathematical notation and R code.</p>
<div id="indexing" class="section level4">
<h4><span class="header-section-number">0.8.0.1</span> Indexing</h4>
<p>Those of us dealing with data almost always have a series of numbers. To describe the concepts in an abstract way, we use indexing. For example 5 numbers:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="dv">1</span><span class="op">:</span><span class="dv">5</span></code></pre></div>
<p>can be generally represented like this <span class="math inline">\(x_1, x_2, x_3, x_4, x_5\)</span>. We use dots to simplify this <span class="math inline">\(x_1,\dots,x_5\)</span> and indexing to simplify even more <span class="math inline">\(x_i, i=1,\dots,5\)</span>. If we want to describe a procedure for a list of any size <span class="math inline">\(n\)</span>, we write <span class="math inline">\(x_i, i=1,\dots,n\)</span>.</p>
<p>We sometimes have two indexes. For example, we may have several measurements (blood pressure, weight, height, age, cholesterol level) for 100 individuals. We can then use double indexes: <span class="math inline">\(x_{i,j}, i=1,\dots,100, j=1,\dots,5\)</span>.</p>
</div>
<div id="summation" class="section level4">
<h4><span class="header-section-number">0.8.0.2</span> Summation</h4>
<p>A very common operation in data analysis is to sum several numbers. This comes up, for example, when we compute averages and standard deviations. If we have many numbers, there is a mathematical notation that makes it quite easy to express the following:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">n &lt;-<span class="st"> </span><span class="dv">1000</span>
x &lt;-<span class="st"> </span><span class="dv">1</span><span class="op">:</span>n
S &lt;-<span class="st"> </span><span class="kw">sum</span>(x)</code></pre></div>
<p>and it is the <span class="math inline">\(\sum\)</span> notation (capital S in Greek):</p>
<p><span class="math display">\[
S = \sum_{i=1}^n x_i
\]</span></p>
<p>Note that we make use of indexing as well. We will see that what is included inside the summation can become quite complicated. However, the summation part should not confuse you as it is a simple operation.</p>
</div>
<div id="greek-letters" class="section level4">
<h4><span class="header-section-number">0.8.0.3</span> Greek letters</h4>
<p>We would prefer to avoid Greek letters, but they are ubiquitous in the statistical literature so we want you to become used to them. They are mainly used to distinguish the unknown from the observed. Suppose we want to find out the average height of a population and we take a sample of 1,000 people to estimate this. The unknown average we want to estimate is often denoted with <span class="math inline">\(\mu\)</span>, the Greek letter for m (m is for mean). The standard deviation is often denoted with <span class="math inline">\(\sigma\)</span>, the Greek letter for s. Measurement error or other unexplained random variability is typically denoted with <span class="math inline">\(\varepsilon\)</span>, the Greek letter for e. Effect sizes, for example the effect of a diet on weight, are typically denoted with <span class="math inline">\(\beta\)</span>. We may use other Greek letters but those are the most commonly used.</p>
<p>You should get used to these four Greek letters as you will be seeing them often: <span class="math inline">\(\mu\)</span>, <span class="math inline">\(\sigma\)</span>, <span class="math inline">\(\beta\)</span> and <span class="math inline">\(\varepsilon\)</span>.</p>
<p>Note that indexing is sometimes used in conjunction with Greek letters to denote different groups. For example, if we have one set of numbers denoted with <span class="math inline">\(x\)</span> and another with <span class="math inline">\(y\)</span> we may use <span class="math inline">\(\mu_x\)</span> and <span class="math inline">\(\mu_y\)</span> to denote their averages.</p>
</div>
<div id="infinity" class="section level4">
<h4><span class="header-section-number">0.8.0.4</span> Infinity</h4>
<p>In the text we often talk about <em>asymptotic</em> results. Typically, this refers to an approximation that gets better and better as the number of data points we consider gets larger and larger, with perfect approximations occurring when the number of data points is <span class="math inline">\(\infty\)</span>. In practice, there is no such thing as <span class="math inline">\(\infty\)</span>, but it is a convenient concept to understand. One way to think about asymptotic results is as results that become better and better as some number increases and we can pick a number so that a computer can’t tell the difference between the approximation and the real number. Here is a very simple example that approximates 1/3 with decimals:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">onethird &lt;-<span class="st"> </span><span class="cf">function</span>(n) <span class="kw">sum</span>( <span class="dv">3</span><span class="op">/</span><span class="dv">10</span><span class="op">^</span><span class="kw">c</span>(<span class="dv">1</span><span class="op">:</span>n))
<span class="dv">1</span><span class="op">/</span><span class="dv">3</span> <span class="op">-</span><span class="st"> </span><span class="kw">onethird</span>(<span class="dv">4</span>)</code></pre></div>
<pre><code>## [1] 3.333e-05</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="dv">1</span><span class="op">/</span><span class="dv">3</span> <span class="op">-</span><span class="st"> </span><span class="kw">onethird</span>(<span class="dv">10</span>)</code></pre></div>
<pre><code>## [1] 3.333e-11</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="dv">1</span><span class="op">/</span><span class="dv">3</span> <span class="op">-</span><span class="st"> </span><span class="kw">onethird</span>(<span class="dv">16</span>)</code></pre></div>
<pre><code>## [1] 0</code></pre>
<p>In the example above, 16 is practically <span class="math inline">\(\infty\)</span>.</p>
</div>
<div id="integrals" class="section level4">
<h4><span class="header-section-number">0.8.0.5</span> Integrals</h4>
<p>We only use these a couple of times so you can skip this section if you prefer. However, integrals are actually much simpler to understand than perhaps you realize.</p>
<p>For certain statistical operations, we need to figure out areas under the curve. For example, for a function <span class="math inline">\(f(x)\)</span> …</p>
<div class="figure"><span id="fig:intergrals"></span>
<img src="bookdown_files/figure-html/intergrals-1.png" alt="Integral of a function." width="672" />
<p class="caption">
图 .: Integral of a function.
</p>
</div>
<p>…we need to know what proportion of the total area under the curve is grey.</p>
<p>The grey area can be thought of as many small grey bars stacked next to each other. The area is then just the sum of the areas of these little bars. The problem is that we can’t do this for every number between 2 and 4 because there are an infinite number. The integral is the mathematical solution to this problem. In this case, the total area is 1 so the answer to what proportion is grey is the following integral:</p>
<p><span class="math display">\[
\int_2^4 f(x) \, dx
\]</span></p>
<p>Because we constructed this example, we know that the grey area is 2.27% of the total. Note that this is very well approximated by an actual sum of little bars:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">width &lt;-<span class="st"> </span><span class="fl">0.01</span>
x &lt;-<span class="st"> </span><span class="kw">seq</span>(<span class="dv">2</span>,<span class="dv">4</span>,width)
areaofbars &lt;-<span class="st">  </span><span class="kw">f</span>(x)<span class="op">*</span>width
<span class="kw">sum</span>( areaofbars )</code></pre></div>
<pre><code>## [1] 0.02299</code></pre>
<p>The smaller we make <code>width</code>, the closer the sum gets to the integral, which is equal to the area.</p>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="index.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="inference.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook/js/app.min.js"></script>
<script src="libs/gitbook/js/lunr.js"></script>
<script src="libs/gitbook/js/plugin-search.js"></script>
<script src="libs/gitbook/js/plugin-sharing.js"></script>
<script src="libs/gitbook/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook/js/plugin-bookdown.js"></script>
<script src="libs/gitbook/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": true,
"facebook": false,
"twitter": true,
"google": false,
"weibo": false,
"instapper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/xie186/DataAnalysisForLifeScience_cn/edit/master/00-author.Rmd",
"text": "编辑"
},
"download": ["bookdown.pdf", "bookdown.epub"],
"toc": {
"collapse": "none"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.bootcss.com/mathjax/2.7.1/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:" && /^https?:/.test(script.src))
      script.src  = script.src.replace(/^https?:/, '');
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
